include/deal.II/base/function.h

```diff
@@ -1148,6 +1148,64 @@ class VectorFunctionFromTensorFunction : public Function<dim, RangeNumberType>
     const std::vector<Point<dim>> &       points,
     std::vector<Vector<RangeNumberType>> &value_list) const override;
 
+  /**
+   * Return the gradient of the specified component of the function at the given
+   * point.
+   */
+  virtual Tensor<1, dim, RangeNumberType>
+  gradient(const Point<dim> & p,
+           const unsigned int component = 0) const override;
+
+  /**
+   * Return the gradient of all components of the function at the given point.
+   */
+  virtual void
+  vector_gradient(
+    const Point<dim> &                            p,
+    std::vector<Tensor<1, dim, RangeNumberType>> &gradients) const override;
+
+  /**
+   * Set <tt>gradients</tt> to the gradients of the specified component of the
+   * function at the <tt>points</tt>.  It is assumed that <tt>gradients</tt>
+   * already has the right size, i.e.  the same size as the <tt>points</tt>
+   * array.
+   */
+  virtual void
+  gradient_list(const std::vector<Point<dim>> &               points,
+                std::vector<Tensor<1, dim, RangeNumberType>> &gradients,
+                const unsigned int component = 0) const override;
+
+  /**
+   * For each component of the function, fill a vector of gradient values, one
+   * for each point.
+   *
+   * The default implementation of this function in Function calls
+   * value_list() for each component. In order to improve performance, this
+   * can be reimplemented in derived classes to speed up performance.
+   */
+  virtual void
+  vector_gradients(const std::vector<Point<dim>> &points,
+                   std::vector<std::vector<Tensor<1, dim, RangeNumberType>>>
+                     &gradients) const override;
+
+  /**
+   * Set <tt>gradients</tt> to the gradients of the function at the
+   * <tt>points</tt>, for all components. It is assumed that
+   * <tt>gradients</tt> already has the right size, i.e. the same size as the
+   * <tt>points</tt> array.
+   *
+   * The outer loop over <tt>gradients</tt> is over the points in the list,
+   * the inner loop over the different components of the function.
+   */
+  virtual void
+  vector_gradient_list(const std::vector<Point<dim>> &points,
+                       std::vector<std::vector<Tensor<1, dim, RangeNumberType>>>
+                         &gradients) const override;
+
+  /**
```

Comment:
I think you don't want to include this comment here.

Reply:
I found a test at base/functions_08.cc. 
I believe I should figure how that works and create something similar. Right? 
I'll take the help of course. How can I build and run one single test?

Reply:
Yes, make a copy of an existing test. You can follow the guide at https://www.dealii.org/developer/developers/testsuite.html#layoutaddtests
This page has a lot of other information to get you started.

Reply:
I made a test called functions_08_gradient based on functions_08. 

0) These test now look a lot like one another. Is it a good/bad idea to make them into one? 
1) Is this good enough or should I make it better? 
2) Does the cli get triggered everytime I push a commit here? Should I be careful about that?  
3) Once this gets merged with master deal, do all the commits that I did show up in the master branch or will my changes show as a one big commit?
4) Some of the pull-requests are getting tags (infrastructure, build system, bug). Is it me who should set that?